File: solutions/zig/sieve/1/sieve.zig
const std = @import("std");

pub fn primes(buffer: []u32, limit: u32) []u32 {
    if (buffer.len == 0 or limit < 2) return buffer[0..0];

    var sieve = [_]bool{true} ** 1024;
    const max = @min(limit, 1023);
    
    sieve[0] = false;
    sieve[1] = false;

    var i: u32 = 2;
    while (i * i <= max) : (i += 1) {
        if (sieve[i]) {
            var j = i * i;
            while (j <= max) : (j += i) {
                sieve[j] = false;
            }
        }
    }

    var count: usize = 0;
    var n: u32 = 0;
    while (n <= max and count < buffer.len) : (n += 1) {
        if (sieve[n]) {
            buffer[count] = n;
            count += 1;
        }
    }

    return buffer[0..count];
}

File: solutions/fortran/sieve/1/sieve.f90
module sieve
  implicit none
  public :: primes

contains
  function primes(limit) result(array)
    integer, intent(in) :: limit
    integer, allocatable :: array(:)
    logical, allocatable :: is_prime(:)
    integer :: i, j, count

    if (limit < 2) then
      allocate(array(0))
      return
    end if

    allocate(is_prime(2:limit))
    is_prime = .true.

    do i = 2, int(sqrt(real(limit)))
      if (is_prime(i)) then
        do j = i*i, limit, i
          is_prime(j) = .false.
        end do
      end if
    end do

    count = 0
    do i = 2, limit
      if (is_prime(i)) then
        count = count + 1
      end if
    end do

    allocate(array(count))

    count = 0
    do i = 2, limit
      if (is_prime(i)) then
        count = count + 1
        array(count) = i
      end if
    end do
  end function primes
end module sieve